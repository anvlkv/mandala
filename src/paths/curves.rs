use cfg_if::cfg_if;

use crate::{GlVec, Point, VectorValuedFn};

/// Quadratic Bezier curve with one control point
#[derive(Debug, Clone, Copy)]
pub struct QuadraticCurve {
    pub start: Point,
    pub control: Point,
    pub end: Point,
}

impl Default for QuadraticCurve {
    fn default() -> Self {
        Self {
            start: GlVec::default().into(),
            control: GlVec::default().into(),
            end: GlVec::default().into(),
        }
    }
}

impl VectorValuedFn for QuadraticCurve {
    fn eval(&self, t: crate::Float) -> crate::Vector {
        crate::Vector {
            x: (1.0 - t).powi(2) * self.start.x
                + 2.0 * (1.0 - t) * t * self.control.x
                + t.powi(2) * self.end.x,
            y: (1.0 - t).powi(2) * self.start.y
                + 2.0 * (1.0 - t) * t * self.control.y
                + t.powi(2) * self.end.y,
            #[cfg(feature = "3d")]
            z: (1.0 - t).powi(2) * self.start.z
                + 2.0 * (1.0 - t) * t * self.control.z
                + t.powi(2) * self.end.z,
        }
    }

    fn length(&self) -> crate::Float {
        let mut length = 0.0;
        let num_segments = 100;
        for i in 0..num_segments {
            let t1 = i as crate::Float / num_segments as crate::Float;
            let t2 = (i + 1) as crate::Float / num_segments as crate::Float;
            let p1 = self.eval(t1);
            let p2 = self.eval(t2);
            length += (p2.x - p1.x).hypot(p2.y - p1.y);
            cfg_if! { if #[cfg(feature = "3d")] {
                length += (p2.z - p1.z).abs();
            }}
        }
        length
    }
}

/// Cubic Bezier curve with two control points
#[derive(Debug, Clone, Copy)]
pub struct CubicCurve {
    pub start: Point,
    pub control1: Point,
    pub control2: Point,
    pub end: Point,
}

impl Default for CubicCurve {
    fn default() -> Self {
        Self {
            start: GlVec::default().into(),
            control1: GlVec::default().into(),
            control2: GlVec::default().into(),
            end: GlVec::default().into(),
        }
    }
}

impl VectorValuedFn for CubicCurve {
    fn eval(&self, t: crate::Float) -> crate::Vector {
        crate::Vector {
            x: (1.0 - t).powi(3) * self.start.x
                + 3.0 * (1.0 - t).powi(2) * t * self.control1.x
                + 3.0 * (1.0 - t) * t.powi(2) * self.control2.x
                + t.powi(3) * self.end.x,
            y: (1.0 - t).powi(3) * self.start.y
                + 3.0 * (1.0 - t).powi(2) * t * self.control1.y
                + 3.0 * (1.0 - t) * t.powi(2) * self.control2.y
                + t.powi(3) * self.end.y,
            #[cfg(feature = "3d")]
            z: (1.0 - t).powi(3) * self.start.z
                + 3.0 * (1.0 - t).powi(2) * t * self.control1.z
                + 3.0 * (1.0 - t) * t.powi(2) * self.control2.z
                + t.powi(3) * self.end.z,
        }
    }

    fn length(&self) -> crate::Float {
        let mut length = 0.0;
        let num_segments = 100;
        for i in 0..num_segments {
            let t1 = i as crate::Float / num_segments as crate::Float;
            let t2 = (i + 1) as crate::Float / num_segments as crate::Float;
            let p1 = self.eval(t1);
            let p2 = self.eval(t2);
            length += (p2.x - p1.x).hypot(p2.y - p1.y);
            cfg_if! { if #[cfg(feature = "3d")] {
                length += (p2.z - p1.z).abs();
            }}
        }
        length
    }
}

#[cfg(test)]
mod curve_tests {
    use super::*;
    use crate::test_util::test_name;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_quadratic_curve_eval() {
        let curve = QuadraticCurve {
            start: Point {
                x: 0.0,
                y: 0.0,
                #[cfg(feature = "3d")]
                z: 0.0,
            },
            control: Point {
                x: 1.0,
                y: 1.0,
                #[cfg(feature = "3d")]
                z: 1.0,
            },
            end: Point {
                x: 2.0,
                y: 2.0,
                #[cfg(feature = "3d")]
                z: 2.0,
            },
        };
        assert_debug_snapshot!(
            test_name("QuadraticCurve"),
            [
                curve.eval(0.0),
                curve.eval(0.25),
                curve.eval(0.5),
                curve.eval(0.75),
                curve.eval(1.0),
            ]
        );
    }

    #[test]
    fn test_cubic_curve_eval() {
        let curve = CubicCurve {
            start: Point {
                x: 0.0,
                y: 0.0,
                #[cfg(feature = "3d")]
                z: 0.0,
            },
            control1: Point {
                x: 1.0,
                y: 1.0,
                #[cfg(feature = "3d")]
                z: 1.0,
            },
            control2: Point {
                x: 2.0,
                y: 2.0,
                #[cfg(feature = "3d")]
                z: 2.0,
            },
            end: Point {
                x: 3.0,
                y: 3.0,
                #[cfg(feature = "3d")]
                z: 3.0,
            },
        };
        assert_debug_snapshot!(
            test_name("CubicCurve"),
            [
                curve.eval(0.0),
                curve.eval(0.25),
                curve.eval(0.5),
                curve.eval(0.75),
                curve.eval(1.0),
            ]
        );
    }
}
