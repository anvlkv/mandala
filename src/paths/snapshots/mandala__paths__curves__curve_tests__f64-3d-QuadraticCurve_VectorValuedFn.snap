---
source: src/paths/curves.rs
expression: "(eval_points, sample_points, derivative_points, normal_points)"
---
(
    [
        Vector3 {
            x: 0.0,
            y: 0.0,
            z: 0.0,
        },
        Vector3 {
            x: 0.2,
            y: 1.8200000000000003,
            z: 0.2,
        },
        Vector3 {
            x: 0.4000000000000001,
            y: 3.2800000000000007,
            z: 0.4000000000000001,
        },
        Vector3 {
            x: 0.6,
            y: 4.38,
            z: 0.6,
        },
        Vector3 {
            x: 0.8,
            y: 5.12,
            z: 0.8,
        },
        Vector3 {
            x: 1.0,
            y: 5.5,
            z: 1.0,
        },
        Vector3 {
            x: 1.2,
            y: 5.52,
            z: 1.2,
        },
        Vector3 {
            x: 1.4,
            y: 5.18,
            z: 1.4,
        },
        Vector3 {
            x: 1.6,
            y: 4.4799999999999995,
            z: 1.6,
        },
        Vector3 {
            x: 1.8,
            y: 3.42,
            z: 1.8,
        },
        Vector3 {
            x: 2.0,
            y: 2.0,
            z: 2.0,
        },
    ],
    [
        Vector3 {
            x: 0.0,
            y: 0.0,
            z: 0.0,
        },
        Vector3 {
            x: 0.2222222222222222,
            y: 2.0,
            z: 0.2222222222222222,
        },
        Vector3 {
            x: 0.4444444444444444,
            y: 3.5555555555555554,
            z: 0.4444444444444444,
        },
        Vector3 {
            x: 0.6666666666666667,
            y: 4.666666666666667,
            z: 0.6666666666666667,
        },
        Vector3 {
            x: 0.8888888888888888,
            y: 5.333333333333333,
            z: 0.8888888888888888,
        },
        Vector3 {
            x: 1.1111111111111112,
            y: 5.555555555555555,
            z: 1.1111111111111112,
        },
        Vector3 {
            x: 1.3333333333333333,
            y: 5.333333333333334,
            z: 1.3333333333333333,
        },
        Vector3 {
            x: 1.5555555555555556,
            y: 4.666666666666666,
            z: 1.5555555555555556,
        },
        Vector3 {
            x: 1.7777777777777777,
            y: 3.5555555555555562,
            z: 1.7777777777777777,
        },
        Vector3 {
            x: 2.0,
            y: 2.0,
            z: 2.0,
        },
    ],
    [
        Vector3 {
            x: 2.0,
            y: 20.0,
            z: 2.0,
        },
        Vector3 {
            x: 2.0625,
            y: 16.0,
            z: 2.0625,
        },
        Vector3 {
            x: 1.875,
            y: 12.0,
            z: 1.875,
        },
        Vector3 {
            x: 2.25,
            y: 10.0,
            z: 2.25,
        },
        Vector3 {
            x: 2.0,
            y: 6.0,
            z: 2.0,
        },
        Vector3 {
            x: 2.0,
            y: 0.0,
            z: 2.0,
        },
        Vector3 {
            x: 2.0,
            y: -2.0,
            z: 2.0,
        },
        Vector3 {
            x: 2.0,
            y: -6.0,
            z: 2.0,
        },
        Vector3 {
            x: 2.0,
            y: -10.0,
            z: 2.0,
        },
        Vector3 {
            x: 1.5,
            y: -13.0,
            z: 1.5,
        },
        Vector3 {
            x: 2.0,
            y: -16.0,
            z: 2.0,
        },
    ],
    [
        Vector3 {
            x: -0.08920645997052801,
            y: 0.10793540029472004,
            z: -0.9901475429766744,
        },
        Vector3 {
            x: -0.11071905166415218,
            y: 0.14108856890839527,
            z: -0.983786210170671,
        },
        Vector3 {
            x: -0.12925535947769864,
            y: 0.17276569934272878,
            z: -0.9764451163157655,
        },
        Vector3 {
            x: -0.16824115042038126,
            y: 0.2522615536871944,
            z: -0.952921310411594,
        },
        Vector3 {
            x: -0.2095465966266709,
            y: 0.3713602101199873,
            z: -0.9045340337332909,
        },
        Vector3 {
            x: -0.0,
            y: 1.0,
            z: -0.0,
        },
        Vector3 {
            x: 0.2113248654051872,
            y: 0.7886751345948129,
            z: 0.5773502691896258,
        },
        Vector3 {
            x: 0.2095465966266709,
            y: 0.3713602101199873,
            z: 0.9045340337332909,
        },
        Vector3 {
            x: 0.1552980596673317,
            y: 0.2235097016633414,
            z: 0.9622504486493764,
        },
        Vector3 {
            x: 0.1009014328020246,
            y: 0.12552091571578694,
            z: 0.9869465034014603,
        },
        Vector3 {
            x: 0.10792720111023596,
            y: 0.13658239111811232,
            z: 0.9847319278346618,
        },
    ],
)
